components/application-mgt/org.wso2.carbon.identity.application.mgt/src/test/java/org/wso2/carbon/identity/application/mgt/DefaultRoleManagementListenerTest.java

```diff
@@ -0,0 +1,224 @@
+import static org.mockito.Mockito.spy;
+import static org.mockito.Mockito.when;
+
+public class DefaultRoleManagementListenerTest {
+
+    private DefaultRoleManagementListener defaultRoleManagementListener;
+    private static final String ROLE_NAME = "test_role";
+    private static final String APPLICATION_NAME = "app_name";
+    private static final String APPLICATION_RES_ID = "app_id";
+    private static final String IS_FRAGMENT_APP = "isFragmentApp";
+    private static final String APPLICATION_AUD = "APPLICATION";
+    private static final String ORGANIZATION_AUD = "ORGANIZATION";
+    private static final String TENANT_DOMAIN = "wso2.com";
+
+    @BeforeClass
+    public void setUp() throws Exception {
+
+        defaultRoleManagementListener = spy(new DefaultRoleManagementListener());
+    }
+
+    @DataProvider(name = "fragmentAppPropertyProvider")
+    public Object[][] fragmentAppPropertyProvider() {
+
+        return new Object[][] {
+                {true},
+                {false}
+        };
+    }
+
+    @Test(dataProvider = "fragmentAppPropertyProvider")
+    public void testPreAddRoleForFragmentApp(boolean isFragmentApp) throws Exception {
+
+        try (MockedStatic<ApplicationManagementService> applicationManagementServiceMockedStatic =
+                     mockStatic(ApplicationManagementService.class)) {
+
+            // Creating service provider object.
+            ServiceProvider serviceProvider = new ServiceProvider();
+            serviceProvider.setApplicationName(APPLICATION_NAME);
+            serviceProvider.setApplicationResourceId(APPLICATION_RES_ID);
+
+            // Creating service provider property object.
+            if (isFragmentApp) {
+                ServiceProviderProperty isFragmentAppSpProp = new ServiceProviderProperty();
+                isFragmentAppSpProp.setName(IS_FRAGMENT_APP);
+                isFragmentAppSpProp.setValue(Boolean.TRUE.toString());
+                serviceProvider.setSpProperties(new ServiceProviderProperty[]{isFragmentAppSpProp});
+            } else {
+                serviceProvider.setSpProperties(null);
+            }
+
+            ApplicationManagementService applicationManagementService = mock(ApplicationManagementService.class);
+            applicationManagementServiceMockedStatic.when(ApplicationManagementService::getInstance).
+                    thenReturn(applicationManagementService);
+            when(applicationManagementService.getApplicationByResourceId(anyString(), anyString())).
+                    thenReturn(serviceProvider);
+            when(applicationManagementService.getAllowedAudienceForRoleAssociation(anyString(), anyString())).
+                    thenReturn(APPLICATION_AUD);
+
+            // Calling the preAddRole method.
+            defaultRoleManagementListener.preAddRole(ROLE_NAME, new ArrayList<>(), new ArrayList<>(),
+                    new ArrayList<>(), APPLICATION_AUD, APPLICATION_RES_ID, TENANT_DOMAIN);
+        }
```

Comment:
At the end can't we validate whether the threadlocal property is set / not according to the scenario

Reply:
Validated